[web-animations] add basic interpolation support for <length> custom properties

graouts opened the pull request:
#### bcd8cc0c0c83b0f2ddb78977a843650168bb138f
<pre>
[web-animations] add basic interpolation support for &lt;length&gt; custom properties
<a href="https://bugs.webkit.org/show_bug.cgi?id=249312">https://bugs.webkit.org/show_bug.cgi?id=249312</a>

Reviewed by Antti Koivisto.

Add very basic interpolation support for custom properties starting with only with
the &lt;length&gt; type and only dealing with non-additive and non-accumulative animations
with explicit keyframe values.

We add a new WPT test for &lt;length&gt; that draws the blueprint for the remaining work.

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-important-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/animation/custom-property-animation-length-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/animation/custom-property-animation-length.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/font-size-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/registered-property-revert-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/resources/utils.js:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::blendSyntaxValues):
(WebCore::blendedCSSCustomPropertyValue):
(WebCore::CSSPropertyAnimation::blendCustomProperty):
* Source/WebCore/animation/CSSPropertyAnimation.h:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::setAnimatedPropertiesInStyle):

Canonical link: <a href="https://commits.webkit.org/257906@main">https://commits.webkit.org/257906@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9d81f16fcbc0786ce0c70503c59935e9a1d00261

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100371 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9533 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33434 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109688 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169922 "Failed to checkout and rebase branch from PR 7615") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104364 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10445 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/80 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92773 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107566 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106149 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7908 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/91169 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/34553 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/22561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77512 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3274 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/24078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3267 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9388 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/43568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5083 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2811 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/animation/CSSPropertyAnimation.h:
```diff
@@ -50,7 +50,7 @@ class CSSPropertyAnimation {
 
     static void blendProperties(const CSSPropertyBlendingClient*, CSSPropertyID, RenderStyle& destination, const RenderStyle& from, const RenderStyle& to, double progress, CompositeOperation, IterationCompositeOperation = IterationCompositeOperation::Replace, double currentIteration = 0);
-    static void blendCustomProperty(const AtomString&, RenderStyle& destination, const RenderStyle& from, const RenderStyle& to, double progress);
+    static void blendCustomProperty(const CSSPropertyBlendingClient*, const AtomString&, RenderStyle& destination, const RenderStyle& from, const RenderStyle& to, double progress, CompositeOperation, IterationCompositeOperation = IterationCompositeOperation::Replace, double currentIteration = 0);
```
Can CSSPropertyBlendingClient be a reference? All callers are providing it.

graouts replied:
Good point. I'll fix this for the new call methods and do a refactor for all the other places in https://bugs.webkit.org/show_bug.cgi?id=249320.

anttijk commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -3910,13 +3911,42 @@ void CSSPropertyAnimation::blendProperties(const CSSPropertyBlendingClient* clie
+{
+    CSSPropertyBlendingContext blendingContext { progress, false, compositeOperation, client, iterationCompositeOperation, currentIteration };
+
+    auto blendInCustomPropertyValueMap = [&](const CustomPropertyValueMap& (RenderStyle::*map)() const, void (RenderStyle::*setValue)(const AtomString&, Ref<CSSCustomPropertyValue>&&)) {
```
Those parameters are nuts. Can you just use `auto` for them? Or not be generic, there are only two cases ever here.

anttijk commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -3910,13 +3911,42 @@ void CSSPropertyAnimation::blendProperties(const CSSPropertyBlendingClient* clie
+        ASSERT(typedCustomPropertyValue<Length>(to.value()));
+        auto blendedLength = blendFunc(*possibleFromLength, *typedCustomPropertyValue<Length>(to.value()), blendingContext);
+        return CSSCustomPropertyValue::createForLengthSyntax(from.name(), WTFMove(blendedLength));
+    }
```
This seems overly complex and hard-to-follow way to do this. I think you should first test if `to` and `from` both contain `CSSCustomPropertyValue::SyntaxValue`. If so it would next invoke another blending function that takes `from` and `to` as `SyntaxValue`s and returns a  `SyntaxValue`. That function then does `switchOn` to the actual type and invokes the type specific being function.

graouts replied:
I attempted to do this in the new patch. Hopefully this is an improvement.

graouts committed:
[web-animations] add basic interpolation support for <length> custom properties
https://bugs.webkit.org/show_bug.cgi?id=249312

Reviewed by Antti Koivisto.

Add very basic interpolation support for custom properties starting with only with
the <length> type and only dealing with non-additive and non-accumulative animations
with explicit keyframe values.

We add a new WPT test for <length> that draws the blueprint for the remaining work.

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-important-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/animation/custom-property-animation-length-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/animation/custom-property-animation-length.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/font-size-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/registered-property-revert-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/resources/utils.js:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::blendSyntaxValues):
(WebCore::blendedCSSCustomPropertyValue):
(WebCore::CSSPropertyAnimation::blendCustomProperty):
* Source/WebCore/animation/CSSPropertyAnimation.h:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::setAnimatedPropertiesInStyle):

Canonical link: https://commits.webkit.org/257906@main
